notion-core/src/main/scala/com/ambiata/notion/core/LocationIO.scala

```diff
@@ -76,6 +76,26 @@ case class LocationIO(configuration: Configuration, s3Client: AmazonS3Client) {
         Hdfs.readWith(new Path(path), f).run(configuration)
     }
 
+  /** write to a given location, using an OutputStream */
+  def writeUnsafe(location: Location)(f: java.io.OutputStream => RIO[Unit]): RIO[Unit] =
+    location match {
+      case l @ LocalLocation(_) =>
+        RIO.safe(l.dirPath.dirname.toFile.mkdirs) >>
+        RIO.using(l.filePath.toOutputStream)(f)
+
+      case HdfsLocation(path) =>
+        Hdfs.writeWith(new Path(path), f).run(configuration)
+
+      case S3Location(bucket, key) =>
+        for {
+          tmpFile <- LocalTemporary.random.fileWithParent
+          out     <- tmpFile.toOutputStream
+          _       <- RIO.using(RIO.ok(out))(f)
+          in      <- tmpFile.toInputStream
+          _       <- RIO.using(RIO.ok(in)) { i => S3Address(bucket, key).putStream(i).execute(s3Client) }
```

Comment:
@etorreborre, sorry but this one has a bug. The reason it has to go to a local file first is to calculate the size, if you don't set the size putStream will have to read everything into memory and will blow up at some point. The easy fix, is to replace the putStream request with a putFile request which will do this check for you.